### PR TITLE
Add USDError class

### PR DESCRIPTION
**Pull request by adlarkin**
Signed-off-by: Ashton Larkin <42042756+adlarkin@users.noreply.github.com>

# 🎉 New feature

## Summary
Adding USD-specific error codes that can be used in the USD component: see https://github.com/ignitionrobotics/sdformat/pull/828#discussion_r792243365

Once this is merged, I will update #828 to use these error codes.

## Test it
N/A

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by adlarkin** [comment]
ABI checker is failing, but I don't think it's because of the changes in this PR:

abi_failure_sdf_usd_error_codes

**Comment by ahcorde** [comment]
These errors are just part of the USD component, does it make sense to include them in the module and not in the hole library ?

What do you think @adlarkin @azeey @scpeters ?

**Comment by adlarkin** [comment]
From the discussion in https://github.com/ignitionrobotics/sdformat/pull/828#discussion_r792243365, I thought it would be okay to add USD error codes to the sdf [ErrorCode enum](https://github.com/ignitionrobotics/sdformat/blob/5cb045c55685c1f98b2026450349b60b44f5682d/include/sdf/Error.hh#L40-L46). While I understand the motivation for placing these error codes in the USD component, I think the issue with that approach is that we'd miss out on the functionality provided by the 
[Error class](https://github.com/ignitionrobotics/sdformat/blob/5cb045c55685c1f98b2026450349b60b44f5682d/include/sdf/Error.hh#L164) (`Error` uses `ErrorCode`).

**Comment by ahcorde** [comment]
@scpeters and @azeey what do you think ?

**Comment by azeey** [comment]
It would nice to separate the USD error codes and put them in the USD component, but I'm not sure know how we can do that and still use `sdf::Error`, so I think this PR is the best approach we have.

**Comment by koonpeng** [comment]
> It would nice to separate the USD error codes and put them in the USD component, but I'm not sure know how we can do that and still use sdf::Error, so I think this PR is the best approach we have.

How about wrapping a usd specific error class in a new `WRAPPED_ERROR` code? A new method will be added to `Error` to retrieved the wrapped error, in order to access the original error object, you will need to do some runtime type introspection.

But then looking at some of the methods in `sdf::Error` like `XmlPath` and `LineNumber`, it seems quite specific to sdf, so if possible, I would prefer using a different class specific to usd errors.

**Comment by adlarkin** [comment]
> How about wrapping a usd specific error class in a new WRAPPED_ERROR code? A new method will be added to Error to retrieved the wrapped error, in order to access the original error object, you will need to do some runtime type introspection.

I'm not sure if I understand the approach proposed here - is the idea to add a new enum to [ErrorCode](https://github.com/ignitionrobotics/sdformat/blob/ff9d635fcc55be1d7e030eb2a455c8638bcfaf98/include/sdf/Error.hh#L46) that would represent a wrapped error class, and then at runtime, if the error is of this wrapped type, then another method can be called to get the underlying error type? In theory, I think this could work, but I'm not sure if there would be any drawbacks to this approach. @azeey @ahcorde what do you think? Should we stick with https://github.com/ignitionrobotics/sdformat/pull/836/commits/00a3dc6b6ed597f0c49439ba2e80b3e46f0bfae3, where the USD error types are listed directly in `ErrorCode`, or should we try to wrap/hide them somehow? There are several other SDF -> USD PRs that are being blocked by this, so it would be nice to figure out the best path moving forward so that we can get this merged.

**Comment by koonpeng** [comment]
> I'm not sure if I understand the approach proposed here - is the idea to add a new enum to [ErrorCode](https://github.com/ignitionrobotics/sdformat/blob/ff9d635fcc55be1d7e030eb2a455c8638bcfaf98/include/sdf/Error.hh#L46) that would represent a wrapped error class, and then at runtime, if the error is of this wrapped type, then another method can be called to get the underlying error type?

Yeah, thats right. I think that if possible, we should use the usd error directly, and only wrap it in a sdf error when we need to pass it to other components.

**Comment by koonpeng** [comment]
I made `kp/wrap-usd-error` to experiment with wrapping errors. It adds a usd specific `sdf::usd::Error` class, one of the error code is a wrapped sdf error. Currently there is no need to propagate any usd errors to sdf components, but if there is a need, we can consider adding a generic error code to `sdf::Error`.

**Comment by azeey** [comment]
@koonpeng I like the approach you prototyped. We might need to be more careful about namespaces since `Error` and `Errors` are defined in `sdf` and `sdf::usd`. And at first I thought there some code duplication between `sdf::Error` and `sdf::usd::Error` and I wanted to see if there's a way to avoid that by using another `sdf::Error` object inside `sdf::usd::Error` to contain the data (eg. `lineNumber`, `filePath`), but I don't think we can avoid writing the constructors and member functions of `sdf::usd::Error`. Your approach gives more flexibility to the `usd` component and other future components to capture the context they need for their specific error types.

Whether `sdf::Error` needs a generic error code remains to be seen, but it might be a good idea to introduce the error code now so we can use it later without breaking ABI.

**Comment by adlarkin** [comment]
@koonpeng, thanks for the prototype - I agree with what @azeey said, this looks pretty good. Unfortunately, there is some code duplication that probably cannot be avoided, but if we want to separate USD errors from SDF errors, this may be the best approach we have at the moment. Regarding namespaces/potential naming collisions, perhaps we can make the following changes to the prototype in `kp/wrap-usd-error`:
* rename the `Error.hh` file in the usd component to `UsdError.hh`
* rename `sdf::usd::ErrorCode` to `sdf::usd::UsdErrorCode`
* rename `sdf::usd::Error` to `sdf::usd::UsdError`
* rename `sdf::usd::Errors` to `sdf::usd::UsdErrors`

I'm not sure if some of the renaming I proposed above makes sense since these objects are already in the USD namespace (perhaps the `Usd` prefix after the `usd` namespace seems redundant), but I would like to figure out a way to make more distinct use of naming/namespaces. Another option we have is to always use the `usd` namespace when creating errors in the USD component (so, code should always read as `usd::Error` instead of `Error`), but this may be difficult to enforce as the USD component codebase continues to grow (I can imagine a developer and/or code reviewer forgetting to enforce this rule at some point).

**Comment by adlarkin** [comment]
@koonpeng after some discussion that was had offline, we are going to go with your approach of creating an error class in the USD component and wrapping `sdf::Error` if needed. Would you mind either opening your own PR  that targets the `sdf12` branch with your prototype, or pushing a commit with your prototype to this PR? If you open your own PR, I will go ahead and close this one.

**Comment by koonpeng** [comment]
@adlarkin I merged my branch with the suggested changes into this pr. I'm going delete my branch now that it is merged in.

**Comment by ahcorde** [comment]
@koonpeng and @adlarkin
 - I fixed the style to match with other files in the library
 - Added `UsdError_TEST.cc`

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Review by ahcorde — APPROVED** [review]

**Review by adlarkin — COMMENTED** [review]
Thanks for the additions @koonpeng and @ahcorde, this looks good to me. I just have one lingering question below that @scpeters or @azeey may want to respond to before we merge this.

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/20b4d82f5db4142cafaa580f40fa43f6c4fe71a2

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/20b4d82f5db4142cafaa580f40fa43f6c4fe71a2

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/20b4d82f5db4142cafaa580f40fa43f6c4fe71a2

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/20b4d82f5db4142cafaa580f40fa43f6c4fe71a2

**Comment by adlarkin** [comment]
All review feedback has been addressed, but I would like to hear what others think about https://github.com/ignitionrobotics/sdformat/pull/836#discussion_r803910763.

File: usd/src/UsdError.cc
Comment (azeey) [inline usd/src/UsdError.cc:153]: There's a code path through the constructor where someone might accidentally set the code to `SDF_ERROR` without assigning `sdf_error`, so consider checking if `sdf_error` has a value before using it?

File: usd/src/UsdError.cc
Comment (azeey) [inline usd/src/UsdError.cc:39]: nit `sdf_error` -> `sdfError`.

File: include/sdf/Error.hh
Comment (azeey) [reply include/sdf/Error.hh:152]: I was the one who suggested that we add this (https://github.com/ignitionrobotics/sdformat/pull/836#issuecomment-1034079463), but if we're going to assign it a large number, I don't think ABI would be an issue, so perhaps it's better to wait till we need it.

Comment (scpeters) [reply include/sdf/Error.hh:152]: ah I see

we can add enum codes without breaking API, so I think it would be fine to omit this for now

Comment (adlarkin) [reply include/sdf/Error.hh]: I removed this type in https://github.com/ignitionrobotics/sdformat/pull/836/commits/851649b4db395da01b34660369269b980b4ef0d6 since we can add it in later if needed; thanks for the feedback :+1:

File: usd/src/UsdError.cc
Comment (adlarkin) [reply usd/src/UsdError.cc]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/287fa31e0ddbdfcecd3309e108b459b00c21515f

File: usd/src/UsdError.cc
Comment (adlarkin) [reply usd/src/UsdError.cc]: Nice catch! I handled this in https://github.com/ignitionrobotics/sdformat/pull/836/commits/287fa31e0ddbdfcecd3309e108b459b00c21515f. I also documented the constructors to note that if the error code is `SDF_ERROR`, the constructor that takes an `sdf::Error` object should be used.

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: I didn't get any compilation warnings when I removed it, so I took it out in https://github.com/ignitionrobotics/sdformat/pull/836/commits/287fa31e0ddbdfcecd3309e108b459b00c21515f

File: usd/include/sdf/usd/UsdError.hh
Comment (adlarkin) [reply usd/include/sdf/usd/UsdError.hh]: https://github.com/ignitionrobotics/sdformat/pull/836/commits/287fa31e0ddbdfcecd3309e108b459b00c21515f

File: usd/src/UsdError.cc
Comment (adlarkin) [reply usd/src/UsdError.cc:98]: Since the string here is stored in a `std::optional`, it looks like the only option we have is to use [`std::reference_wrapper`](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) ([source](https://stackoverflow.com/questions/63773503/can-i-somehow-return-stdoptional-with-const-reference-to-my-data-to-avoid-copy)). This would mean that users have to make an extra `get()` call in order to get the actual const ref string. You can see how this would look in the tests:

<details>
  <summary>std::reference_wrapper diff</summary>

```diff
diff --git a/usd/include/sdf/usd/UsdError.hh b/usd/include/sdf/usd/UsdError.hh
index 68cc1f4e..10646ee7 100644
--- a/usd/include/sdf/usd/UsdError.hh
+++ b/usd/include/sdf/usd/UsdError.hh
@@ -17,6 +17,7 @@
 #ifndef SDF_USD_USDERROR_HH_
 #define SDF_USD_USDERROR_HH_
 
+#include <functional>
 #include <optional>
 #include <string>
 #include <vector>
@@ -108,7 +109,8 @@ namespace sdf
     /// \brief Get the file path associated with this error.
     /// \return Returns the path of the file that this error is related to,
     /// nullopt otherwise.
-    public: std::optional<std::string> FilePath() const;
+    public: std::optional<std::reference_wrapper<const std::string>>
+            FilePath() const;
 
     /// \brief Sets the file path that is associated with this error.
     /// \param[in] _filePath The file path that is related to this error. (e.g.
diff --git a/usd/src/UsdError.cc b/usd/src/UsdError.cc
index 0d31b94f..2d0f76ff 100644
--- a/usd/src/UsdError.cc
+++ b/usd/src/UsdError.cc
@@ -17,6 +17,7 @@
 
 #include "sdf/usd/UsdError.hh"
 
+#include <functional>
 #include <iostream>
 
 using namespace sdf::usd;
@@ -95,7 +96,8 @@ const std::string &UsdError::Message() const
 }
 
 /////////////////////////////////////////////////
-std::optional<std::string> UsdError::FilePath() const
+std::optional<std::reference_wrapper<const std::string>>
+UsdError::FilePath() const
 {
   return this->dataPtr->filePath;
 }
diff --git a/usd/src/UsdError_TEST.cc b/usd/src/UsdError_TEST.cc
index 9c7a8768..caeee345 100644
--- a/usd/src/UsdError_TEST.cc
+++ b/usd/src/UsdError_TEST.cc
@@ -37,7 +37,7 @@ TEST(Error, DefaultConstruction)
   }
   error.SetFilePath("/usd/world");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ("/usd/world", error.FilePath());
+  EXPECT_EQ("/usd/world", error.FilePath().value().get());
 
   error.SetLineNumber(5);
   ASSERT_TRUE(error.LineNumber().has_value());
@@ -73,7 +73,7 @@ TEST(Error, ValueConstructionWithFile)
   EXPECT_EQ(error.Code(), sdf::usd::UsdErrorCode::INVALID_PRIM_PATH);
   EXPECT_EQ(error.Message(), "Invalid Prim path");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ(error.FilePath().value(), emptyFilePath);
+  EXPECT_EQ(error.FilePath().value().get(), emptyFilePath);
   EXPECT_FALSE(error.LineNumber().has_value());
   EXPECT_FALSE(error.SdfError().has_value());
 
@@ -95,7 +95,7 @@ TEST(Error, ValueConstructionWithLineNumber)
   EXPECT_EQ(error.Code(), sdf::usd::UsdErrorCode::INVALID_PRIM_PATH);
   EXPECT_EQ(error.Message(), "Invalid Prim path");
   ASSERT_TRUE(error.FilePath().has_value());
-  EXPECT_EQ(error.FilePath().value(), emptyFilePath);
+  EXPECT_EQ(error.FilePath().value().get(), emptyFilePath);
   ASSERT_TRUE(error.LineNumber().has_value());
   EXPECT_EQ(error.LineNumber().value(), lineNumber);
   EXPECT_FALSE(error.SdfError().has_value());
```

</details>

Do you think this is a good approach to take? Or should we leave the method signature as-is, returning a `std::optional<std::string>`?

Comment (azeey) [reply usd/src/UsdError.cc:98]: Oh, I meant `const std::optional<std::string> &`. Sorry for not being clear.

File: usd/src/UsdError.cc
Comment (azeey) [reply usd/src/UsdError.cc]: LGTM!

File: usd/src/UsdError.cc
Comment (adlarkin) [reply usd/src/UsdError.cc]: No problem, yeah we can definitely do that - I made the change in https://github.com/ignitionrobotics/sdformat/pull/836/commits/4be0adaa92bc12505a14a647cc2183b0079ec8c3

**Comment by adlarkin** [comment]
I'm planning to merge this if CI comes back green, unless anyone has any other final review comments.

File: usd/include/sdf/usd/UsdError.hh
Comment (azeey) [inline usd/include/sdf/usd/UsdError.hh:97]: nit:
```suggestion
    public: explicit UsdError(const sdf::Error &_sdfError);
```

**Review by azeey — APPROVED** [review]
LGTM!

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**
